drivers: video: Fix blocking `video_buffer_alloc`

epc-ake opened the pull request:
This PR fixes a blocking call to `video_buffer_alloc` in case of memory shortage by introducing a the Kconfig option `CONFIG_VIDEO_BUFFER_ALLOC_TIMEOUT`. 

I've added a new Kconfig option for backwards compatibility, but I'm open to alternative approaches.

ngphibang commented on drivers/video/Kconfig:
```diff
@@ -50,6 +50,10 @@ config VIDEO_BUFFER_SMH_ATTRIBUTE
 	  1: SMH_REG_ATTR_NON_CACHEABLE
 	  2: SMH_REG_ATTR_EXTERNAL
 
+config VIDEO_BUFFER_ALLOC_TIMEOUT
```
I don't see where do you use this new config ? Also, this change should not be in the same commit with the change in esp32 power management.

loicpoulain commented:
Wrong commit content?

epc-ake committed:
drivers: video: Add timeout to`video_buffer_alloc`

This PR fixes a blocking call to video_buffer_alloc in case of memory
shortage by addign a timeout parameter to the API.

Signed-off-by: Armin Kessler <ake@espros.com>

epc-ake commented:
I accidentally committed the wrong file... 
Fixed now.

ngphibang commented:
Wouldn't be it's better to expose the timeout as a parameter in `video_buffer_aligned_alloc()` and `video_buffer_alloc()` ? This way, users can easily / directly specify the timeout as they want as well as do a non-blocking version with K_NO_WAIT ?

It helps to avoid an extra config that 's not easier to discover ?

loicpoulain commented:
> Wouldn't be better to expose the `timeout` as a parameter in `video_buffer_aligned_alloc()` and `video_buffer_alloc()` ? This way, users can easily / directly specify the `timeout` as they want and the `timeout` can also be set **differently** in several places in the code, as well as they can do even a non-blocking version with `K_NO_WAIT` ?
> 
> It also helps to avoid an extra config that 's not easy to be discovered

+1, to me it should not depend on a config, it depends from which context you call that function.

wmrsouza commented:
> Wouldn't be better to expose the `timeout` as a parameter in `video_buffer_aligned_alloc()` and `video_buffer_alloc()` ? This way, users can easily / directly specify the `timeout` as they want and the `timeout` can also be set **differently** in several places in the code, as well as they can do even a non-blocking version with `K_NO_WAIT` ?
> 
> It also helps to avoid an extra config that 's not easy to be discovered

+1

epc-ake commented:
> Wouldn't be better to expose the `timeout` as a parameter in `video_buffer_aligned_alloc()` and `video_buffer_alloc()` ? This way, users can easily / directly specify the `timeout` as they want and the `timeout` can also be set **differently** in several places in the code, as well as they can do even a non-blocking version with `K_NO_WAIT` ?
> 
> It also helps to avoid an extra config that 's not easy to be discovered

Sure, I changed it to this approach.

josuah commented:
There are a few [compliance checks](https://github.com/zephyrproject-rtos/zephyr/actions/runs/11610056805/job/32328583701?pr=80665) that do not pass for it to be merged, but with the new approach, it is easier to handle short memory conditions and to debug.

This will also need to have the samples updated to the new API. To reproduce some of the CI locally:

Compliance check (ignore all CONFIG_LV_ or CONFIG_PERCEPIO_ related errors):
```
python scripts/ci/check_compliance.py --exclude-module ModulesMaintainers
```

Twister build tests:
```
west twister --inline-logs --platform native_sim \
  --scenario tests/drivers/build_all/video/drivers.video.build
```

Doxygen documentation build
```
make -C ./doc doxygen
```

It takes about 1 minute in total, faster than waiting for the CI.
Thank you for this fix! I had it in the back of my mind as well.

uLipe reviewed (commented):
Thee new approach looks good to me, just address  the CI claiming :)

sylvioalves commented on include/zephyr/drivers/video.h:
```diff
@@ -732,19 +732,21 @@ static inline int video_set_signal(const struct device *dev, enum video_endpoint
  *
  * @param size Size of the video buffer (in bytes).
  * @param align Alignment of the requested memory, must be a power of two.
+ * @param timeout How long to wait, or K_NO_WAIT.
```
I suggest rephrasing this to:
`@param timeout Timeout duration in milliseconds or K_NO_WAIT`

josuah replied:
If it is a [`k_timeout_t`](https://docs.zephyrproject.org/apidoc/latest/structk__timeout__t.html), would the unit depend on which macro is used? `K_MSEC()`, `K_USEC()`...

sylvioalves replied:
Oh, true. My intention is to remove the "How long to wait" from there. Agrre that "milliseconds" isn't correct.

ngphibang commented:
I suggest to change the commit title / message as well if possible. It's rather an extension of video_buffer_alloc() to expose the timeout parameter rather than a bug fix.

kartben reviewed (changes requested):
The breaking API change needs to be documented in the migration guide, or you can consider making it non breaking by keeping the old signature/behavior and having its implementation wrap the new API and call it with timeout=K_FOREVER. 
To be clear it's perfectly OK to "break" the API as video is experimental, and whether you want to keep the legacy API or not is really up to you / video collaborators to decide

josuah commented:
Thank you for your guidance about the migration guide!

https://docs.zephyrproject.org/latest/releases/index.html#migration-guides
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-4.1.rst
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-4.1.rst

kartben commented:
@epc-ake will you be able to revisit this PR?

epc-ake commented:
> @epc-ake will you be able to revisit this PR?

I‘ll have a look tomorrow. 
You just want me to add a note to the migration guide for v4.1 and add it to this pr. Right?

kartben commented:
> > @epc-ake will you be able to revisit this PR?
> 
> I‘ll have a look tomorrow. 
> You just want me to add a note to the migration guide for v4.1 and add it to this pr. Right?

Yes, please. Thanks!

kartben commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -49,6 +49,9 @@ LVGL
 Device Drivers and Devicetree
 *****************************
 
+* The :c:func:`video_buffer_alloc` and :c:func:`video_buffer_aligned_alloc` function in the
+  video API now includes a timeout parameter.
```
```suggestion
* The :c:func:`video_buffer_alloc` and :c:func:`video_buffer_aligned_alloc` functions in the
  video API now take an additional timeout parameter.
```
